File: 24.Web-Sockets-And-Real-time-Communication-with-WebSockets/code/typescript/a_websocket_server_1.ts
import { WebSocketServer, WebSocket } from 'ws';
import * as http from 'http';

const server = http.createServer();
const wss = new WebSocketServer({ noServer: true });

server.on('upgrade', (request, socket, head) => {
    // CheckOrigin guards against cross-site hijacking , DO NOT leave it open (sec 14).
    const origin = request.headers.origin;
    if (origin !== "https://app.example.com") {
        socket.write('HTTP/1.1 401 Unauthorized\r\n\r\n');
        socket.destroy();
        return;
    }

    wss.handleUpgrade(request, socket, head, (ws) => {
        // performs the 101 handshake (sec 4)
        wss.emit('connection', ws, request);
    });
});

wss.on('connection', (ws: WebSocket) => {
    // the READ LOOP , one per connection
    ws.on('message', (message: Buffer, isBinary: boolean) => {
        // echo it straight back
        ws.send(message, { binary: isBinary });
    });

    ws.on('close', () => {
        // client closed or connection died (sec 6) -> exit, cleanup
    });
});

server.listen(8080, () => {
    console.log('Listening on port 8080');
});

File: 24.Web-Sockets-And-Real-time-Communication-with-WebSockets/code/rust/a_websocket_server_1.rs
use axum::{
    extract::{
        ws::{Message, WebSocket, WebSocketUpgrade},
        TypedHeader,
    },
    headers,
    response::IntoResponse,
    routing::get,
    Router,
};
use futures::{sink::SinkExt, stream::StreamExt};
use std::net::SocketAddr;

#[tokio::main]
async fn main() {
    let app = Router::new().route("/ws", get(ws_handler));
    
    let addr = SocketAddr::from(([127, 0, 0, 1], 8080));
    axum::Server::bind(&addr)
        .serve(app.into_make_service())
        .await
        .unwrap();
}

async fn ws_handler(
    ws: WebSocketUpgrade,
    // CheckOrigin guards against cross-site hijacking , DO NOT leave it open (sec 14).
    user_agent: Option<TypedHeader<headers::Origin>>,
) -> impl IntoResponse {
    let origin = user_agent.map(|o| o.to_string()).unwrap_or_default();
    
    // In axum, you would typically check this in a middleware, but here is inline:
    if origin != "https://app.example.com" {
        // Return unauthorized or just let it connect and close (omitted for brevity)
    }

    // performs the 101 handshake (sec 4)
    ws.on_upgrade(handle_socket)
}

async fn handle_socket(mut socket: WebSocket) {
    // the READ LOOP , one per connection
    while let Some(msg) = socket.next().await {
        if let Ok(msg) = msg {
            // echo it straight back
            if socket.send(msg).await.is_err() {
                break; // client closed or connection died (sec 6) -> exit, cleanup via drop
            }
        } else {
            break; 
        }
    }
    // socket is dropped and closed automatically (frees the socket)
}
